### PR TITLE
Add support for ios_*_test_suite targets as top level targets

### PR DESCRIPTION
**Pull request by thiagohmcruz**
Resolves https://github.com/MobileNativeFoundation/rules_xcodeproj/issues/1956

- [ ] TBD: Add tests (both for integration and rules_ios)

Initial intend was to add support for `test_suite` rules and map each runner in `runners` to targets in Xcode, this is doable but it creates indexing issues due to two targets sharing paths. Ended up changing the approach to simply skip the runners (like it's done for non-suite rules already) and let Xcode be the runner, only propagating the bundle. This is probably similar to how Tulsi handled this based on the [docs](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-ios.md#ios_unit_test):

>Builds and bundles an iOS Unit .xctest test bundle. Runs the tests using the provided test runner when invoked with bazel test. When using Tulsi to run tests built with this target, runner will not be used since Xcode is the test runner in that case.

The only catch here is that `replacement _labels` will contain many underlying labels associated with the same target ID. So added logic to unify that and bring back a `id <=> label` 1-1 relationship, e.g., if these are found in this id => labels mapping exists in the replacement labels:
```py
{
  "@//tests/macos/xcodeproj:Single-Application-RunnableTestSuite.__internal__.__test_bundle applebin_ios-ios_sim_arm64-dbg-ST-d1716b12dfa6": [
    Label("//tests/macos/xcodeproj:Single-Application-RunnableTestSuite_iPad-Air-2__16.2"), 
    Label("//tests/macos/xcodeproj:Single-Application-RunnableTestSuite_iPhone-13-Pro__16.2")
  ]
}
```

Then the longest common value between the labels and the label contained in the id will be used to bring back a 1-1 relationship:
```py
{
  "@//tests/macos/xcodeproj:Single-Application-RunnableTestSuite.__internal__.__test_bundle applebin_ios-ios_sim_arm64-dbg-ST-d1716b12dfa6": Label("@//tests/macos/xcodeproj:Single-Application-RunnableTestSuite"),
}
```

### CODE REVIEW

**Review by brentleyjones — APPROVED** [review]
Approving to run CI

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 1]
```diff
@@ -71,6 +71,7 @@ _BUILD_TEST_RULES = {
     "macos_build_test": None,
     "tvos_build_test": None,
     "watchos_build_test": None,
+    "test_suite": None,
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:74]: nit: this should only be the `*_build_test` rules.

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 2]
```diff
@@ -259,6 +260,15 @@ def _skip_target(
...
 
+    test_suite_replacement_labels = []
+    if ctx.rule.kind == "test_suite":
+        for info in valid_transitive_infos:
+            for xcode_target in info.xcode_targets.to_list():
+                if xcode_target.id.count(".__internal__.__test_bundle"):
+                    test_suite_replacement_labels.append(
+                        struct(id = xcode_target.id, label = target.label),
+                    )
+
```
Comment (thiagohmcruz) [inline xcodeproj/internal/xcodeprojinfo.bzl:271]: @brentleyjones curious if you see a better approach here. Without this the scheme is Xcode gets named after the latest label from the underlying runners: https://github.com/bazel-ios/rules_ios/blob/master/tests/macos/xcodeproj/BUILD.bazel#L55 (which was the `iPhone` one based on my testing)

But when using a test suite I'd say the expected behaviour is to see the name of the respective test suite in Xcode: https://github.com/bazel-ios/rules_ios/blob/master/tests/macos/xcodeproj/BUILD.bazel#LL41C13-L41C49

All replacements are happening on top of the same label for the internal `.__internal__.__test_bundle` target, so adding this as the last option below as `+ test_suite_replacement_labels` fixed it for me but maybe there's a better approach here

Also this fixed adding the `test_suite` rule as a custom scheme in `schemes`, otherwise if fails since it can't find the `test_suite` label in the transitive deps and finds only the `iPhone/iPad` ones, so It ends up hitting this: https://github.com/MobileNativeFoundation/rules_xcodeproj/blob/main/tools/generator/src/Extensions/XcodeScheme%2BExtensions.swift#L27

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 1 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:74]: 👍 will put this in a separate constant

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 2 above)
Comment (brentleyjones) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: When it's a test suite, I think we shouldn't set `depset(direct)` below in `replacement_labels`, since we want the labels to be left alone, right?

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: >When it's a test suite, I think we shouldn't set depset(direct) below in replacement_labels, since we want the labels to be left alone, right?

Ok, let me look into that!

>I don't think we want the test suite to appear in Xcode, I think the various sub-tests should just be included.

Maybe I'm thinking about this backwards, I thought it made sense to have the test suite there and Xcode is the "runner" but now I think that missed the point of having different runners already configured in `BUILD` files. Ok, let me explore the approach above.

Comment (luispadron) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: > I don't think we want the test suite to appear in Xcode, I think the various sub-tests should just be included.

It could become an auto-generated scheme that builds & tests all the tests under it maybe? But right, I think the default would be that it simply includes whatever test targets are declared in the suite.

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: Yep, having all underlying tests in Xcode seems to be the correct behaviour indeed, discussing possible approaches offline to not pollute this thread too much but the TL;DR is that both underlying tests depend on the same `.__internal__.__test_bundle` test bundle target which is causing the latest one in `replacement_labels` to override all the others and we end up with a single test target in Xcode

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: >It could become an auto-generated scheme that builds & tests all the tests under it maybe

This seems wroth exploring, haven't looked into auto-generated schemes at all but will do

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:271]: Ok, after some discussion I'll remove this and go with a different approach, will detailed it better in the PR descr once it's ready for review

File: xcodeproj/internal/target_id.bzl [hunk 3]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+        The longest common prefix between the list of labels and the label contained in the id,
+        if a longest common prefix is not found the label contained in the id is returned
+    """
+    id_label = id.split(" ")[0]  # This assumes the id follows the creation pattern in `get_id`
+
+    res = _longest_common_prefix(
+        [id_label] + ["%s" % label for label in replacement_labels],
+    )
+    if not res:
+        return None
```
Comment (thiagohmcruz) [inline xcodeproj/internal/target_id.bzl:59]: Should probably `fail` here since ideally this would always return something?

File: xcodeproj/internal/target_id.bzl [hunk 4]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+        for label in labels:
+            if label[index] != c:
+                return res[:index]
+        index += 1
+
+    # If the above never returns the shortest string is the result
+    return res
+
+def calculate_replacement_label(*, id, replacement_labels):
+    """Calculates a single replacement label for a given id and list of replacement labels
```
Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:40]:
```suggestion
    """Calculates a single replacement label for a given id and list of \
    replacement labels
```

Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:40]:
```suggestion
    """Calculates a single replacement label for a given id and list of \
    replacement labels.
```

File: xcodeproj/internal/target_id.bzl [hunk 5]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+
+    It considers the list of labels received plus the label contained in the id itself, then it finds
+    the longest common prefix between those
```
Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:43]:
```suggestion
    It considers the list of labels received plus the label contained in the id
    itself, then it finds the longest common prefix between those.
```

Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:43]:
```suggestion
    It considers the list of labels received plus the label contained in the id
    itself, then it finds the longest common prefix between those.
```

File: xcodeproj/internal/target_id.bzl [hunk 6]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+
+    Args:
+        id: An id that uniquely represents a target (see `get_id`)
+        replacement_labels: A list of replacement labels
+
+    Returns:
+        The longest common prefix between the list of labels and the label contained in the id,
+        if a longest common prefix is not found the label contained in the id is returned
+    """
+    id_label = id.split(" ")[0]  # This assumes the id follows the creation pattern in `get_id`
```
Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:53]: Why do you need to do this part? Can't `_longest_common_prefix` just be called with `replacement_labels`?

Related: the format of a target id is supposed to be opaque, and will probably change in the future to be a  non-identifying hash.

File: xcodeproj/internal/target_id.bzl [hunk 7]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+
+    res = _longest_common_prefix(
+        [id_label] + ["%s" % label for label in replacement_labels],
```
Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:56]:
```suggestion
        [id_label] + [str(label) for label in replacement_labels],
```

Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:56]:
```suggestion
        [id_label] + [bazel_labels.normalize_label(label) for label in replacement_labels],
```

File: xcodeproj/internal/target_id.bzl [hunk 8]
```diff
@@ -12,6 +14,53 @@ def get_id(*, label, configuration):
...
+    )
+    if not res:
+        return None
+
+    res = bazel_labels.normalize_label(res)
```
Comment (brentleyjones) [inline xcodeproj/internal/target_id.bzl:61]: No need to normalize here, as you should do it above (since `bazel_labels.normalize_label` basically does `str()` already, with a little more magic).

File: xcodeproj/internal/xcodeproj_rule.bzl [hunk 9]
```diff
@@ -1485,13 +1485,29 @@ configurations: {}""".format(", ".join(xcode_configurations)))
...
+    # Ensures the id <=> label relationship is 1-1
+    replacement_labels = {}
+    for r in replacement_labels_infos:
+        if r.id in multiple_replacement_labels:
+            replacement_labels[r.id] = calculate_replacement_label(
+                id = r.id,
+                replacement_labels = multiple_replacement_labels[r.id],
+            )
+        else:
+            replacement_labels[r.id] = r.label
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeproj_rule.bzl:1509]: I think it can be expressed more efficiently this way:
```suggestion
    raw_replacement_labels = {}
    for r in replacement_labels_infos:
        raw_replacement_labels.setdefault(r.id, []).append(r.label)

    replacement_labels = {}
    for id, labels in raw_replacement_labels.items():
        if len(labels) > 1:
            replacement_labels[id] = calculate_replacement_label(
                 id = id,
                 replacement_labels = labels,
             )
        else:
            replacement_labels[id] = labels[0]
```

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 10]
```diff
@@ -71,6 +71,11 @@ _BUILD_TEST_RULES = {
     "macos_build_test": None,
     "tvos_build_test": None,
     "watchos_build_test": None,
+    "test_suite": None,
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:74]: Still needs to be removed.

File: xcodeproj/internal/xcodeprojinfo.bzl [hunk 11]
```diff
@@ -88,6 +93,8 @@ def _should_skip_target(*, ctx, target):
     """
     if ctx.rule.kind in _BUILD_TEST_RULES:
         return True
+    if ctx.rule.kind in _TEST_SUITE_RULES:
+        return True
```
Comment (brentleyjones) [inline xcodeproj/internal/xcodeprojinfo.bzl:97]: I think this, and the replacement labels stuff, can be two separate PRs.

- Replacement labels: can add an example to `examples/integrations` for an `ios_unit_test` with multiple runners (or just adjust one of existing ones to have multiple runners).
- `test_suite`: Can add a `test_suite` around two tests in `examples/integration` and list it in `top_level_targets`

File: xcodeproj/internal/target_id.bzl (hunk 6 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/target_id.bzl:53]: good catch, we don't need to pass only the label here, the common prefix will be the same anyways 🤦 

>the format of a target id is supposed to be opaque, and will probably change in the future to be a non-identifying hash.
Ok good to know, I wasn't sure if we wanted to assume anything about that format or not, it makes sense to ensure it's opaque though, will update this code!

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 10 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:74]: Oops

File: xcodeproj/internal/xcodeprojinfo.bzl (hunk 11 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:97]: Ok, after we reverted the approach of bringing all runner variations into Xcode I thought this could go back to being a single change but I could be missing something:

>Replacement labels: can add an example to examples/integrations for an ios_unit_test with multiple runners (or just adjust one of existing ones to have multiple runners).

I think `ios_unit_test` [only takes a single runner](https://github.com/bazelbuild/rules_apple/blob/master/doc/rules-ios.md#ios_unit_test)? 

>test_suite: Can add a test_suite around two tests in examples/integration and list it in top_level_targets

That would be adding `ios_unit_test_suite` and `ios_ui_test_suite` with `runners` set and then `rules_ios`'s `ios_unit_test` with `runners` set (`rules_ios` call the above two under the hood). IIUC this already covers what you intended to cover in the previous item making it unnecessary?

Comment (brentleyjones) [reply xcodeproj/internal/xcodeprojinfo.bzl:97]: Got confused. Can land `test_suites` support first (just raw `test_suite`, not `ios_unit_test_suite`), then this change with the `replacement_labels` fix can come after to add support for `ios_unit_test_suite`/`rules_ios`'s `ios_unit_test.runners`.

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:97]: Ok thx for clarifying, I had only the `ios_*` rules in mind, didn't realize we could (and should) test `test_suite` rules explicitly first. Decoupling things like this makes sense from the perspective of `rules_xcodeproj` regardless how each repo abstracts these, thx for clarifying.

Will split this into the two PRs then!

Comment (luispadron) [reply xcodeproj/internal/xcodeprojinfo.bzl:97]: +1 to supporting test_suite. On Cash we use the native rule along with multiple ios_unit_test and `target_compatible_with` to split up our snapshot tests by device. 

Each unit test target specifies a runner and they're all part of the same suite. Ideally, this would show up as a single target in Xcode where the runner is whatever Xcode has selected. 

Right now I see all 4 targets get loaded and running it spins up the entire test suite, which spins up 4 sims + 1 sim from Xcode

Comment (thiagohmcruz) [reply xcodeproj/internal/xcodeprojinfo.bzl:97]: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2184

Just one Q there about custom schemes

File: xcodeproj/internal/target_id.bzl (hunk 6 above)
Comment (thiagohmcruz) [reply xcodeproj/internal/target_id.bzl:54]: When passing the `id` the results works if the labels are all in the same format, suggesting a fix for that here first: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2188

This way I can compare the labels with the label in the id on the same basis to find the common prefix

**Comment by thiagohmcruz** [comment]
All the discussion and feedback here was very helpful, thx @brentleyjones for the offline suggestions/guidance too!

Closing in favour of: https://github.com/MobileNativeFoundation/rules_xcodeproj/pull/2196
